subprojects/build-option/src/main/java/org/gradle/internal/buildoption/BuildOption.java

```diff
@@ -33,4 +34,15 @@
     void applyFromProperty(Map<String, String> properties, T settings);
     void configure(CommandLineParser parser);
     void applyFromCommandLine(ParsedCommandLine options, T settings);
+
+    enum Origin {
+        GRADLE_PROPERTY, COMMAND_LINE;
+
+        public void handleInvalidValue(String gradlePropertyFailureMessage, String commandLineFailureMessage) {
```

Comment:
This still requires duplicated logic. The method should take the expected and actual value and construct the correct messages itself.

Reply:
At least at the moment we still have a lot of custom messages (see other implementations of `BuildOption`). That's why I didn't move the message creation there yet. I think we can do that in another cleanup PR. I am sure you'll have to touch another 10 or so files.